rapid7/metasploit-framework
# Create exploit for CVE-2016-6563 / Dlink DIR HNAP Login bof

Several Dlink routers contain a pre-authentication stack buffer overflow vulnerability, which is exposed on the LAN interface on port 80. This vulnerability affects the HNAP SOAP protocol, which accepts arbitrarily long strings into certain XML parameters and then copies them into the stack. 
This exploit has been tested on the real devices DIR-818LW and 868L (rev. B), and it was tested using emulation on the DIR-822, 823, 880, 885, 890 and 895. Others might be affected, and this vulnerability is present in both MIPS and ARM devices.
The MIPS devices are powered by Lextra RLX processors, which are crippled MIPS cores lacking a few load and store instructions. Because of this the payloads have to be sent unencoded, which can cause them to fail, although the bind shell seems to work well.
For the ARM devices, the inline reverse tcp seems to work best.
Check the reference links to see the vulnerable firmware versions.

## Verification

I've included the payload status as a comment above, feel free to remove it. On the MIPS versions, the payloads that can be sent with null bytes work well. No encoder can be used for these ones, as the encoder relies in MIPS instructions that the router CPU does not support. So in practice, only really the bind shell and the command payloads work.
As for the ARM payloads, they are tricky. I had many conversations on IRC with some of the devs, and only the inline payloads work well. The others fail constantly.
Please note that these are the tests done on real devices (818 and 868). In qemu most payloads work well.

How to test:
- Connect to the wireless LAN
- Choose the correct target (ARM or MIPS, according to the router model)
- Choose a good payload (inline preferably)
- Fire and get a shell!

A pcap is attached for a test on the qemu emulated 823 model.

[hnap-823-revshell.pcap.zip](https://github.com/rapid7/metasploit-framework/files/576183/hnap-823-revshell.pcap.zip)
